astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        # Then calculate the resulting data (which can but not needs to be a
+        # quantity)
+        result = self._arithmetic_data(operation, operand, **kwds)
+        # Determine the other properties
+        if propagate_uncertainties is None:
+            kwargs['uncertainty'] = None
+        elif not propagate_uncertainties:
+            if self.uncertainty is None:
+                kwargs['uncertainty'] = deepcopy(operand.uncertainty)
+            else:
+                kwargs['uncertainty'] = deepcopy(self.uncertainty)
+        else:
+            kwargs['uncertainty'] = self._arithmetic_uncertainty(
+                operation, operand, result, uncertainty_correlation, **kwds)
+        if handle_mask is None:
+            kwargs['mask'] = None
+        elif not handle_mask:
+            if self.mask is None:
+                kwargs['mask'] = deepcopy(operand.mask)
+            else:
+                kwargs['mask'] = deepcopy(self.mask)
+        else:
+            kwargs['mask'] = self._arithmetic_mask(operand, **kwds)
 
-        from .. import conf
-
-        if self.wcs != operand.wcs:
-            raise ValueError("WCS properties do not match")
-
-        # get a sensible placeholder if .unit is None
-        self_unit = self.unit or dimensionless_unscaled
-        operand_unit = operand.unit or dimensionless_unscaled
-
-        # This check could be rolled into the calculation of the result
-        # but checking now avoids a potentially expensive calculation that
-        # would fail anyway.
-        try:
-            # Quantity is designed to work with numpy ufuncs, but plain
-            # Unit is not, so convert units to quantities
-            result_unit = operation(1 * self_unit, 1 * operand_unit).unit
-        except UnitsError:
-            # current API raises ValueError in this case, not UnitError
-            raise ValueError("operand units do not match")
-
-        if self.data.shape != operand.data.shape:
-            raise ValueError("operand shapes do not match")
-
-        # Instead of manually scaling the operand data just let Quantity
-        # handle it.
-        # Order of the arguments is important here if the operation is
-        # addition or subtraction and the units of the operands are different
-        # but compatible. NDData follows the convention that Quantity follows
-        # in that case, with the units of the first operand (i.e. self)
-        # determining the units of the result.
-        data = operation(self.data * self_unit, operand.data * operand_unit)
-
-        result_unit = data.unit
-        # If neither self nor operand had units then should return a result
-        # that has no unit. A check that the result_unit is dimensionless
-        # should not be necessary, but also does no harm.
-        if self.unit is None and operand.unit is None:
-            if result_unit is dimensionless_unscaled:
-                result_unit = None
+        if handle_meta is None:
```

Comment:
I think similar to `wcs` this maybe should be part of the metaclass - you just do the arithmetic on the meta class and it will know what to do. This would then also be the same for tables, which should show the same behaviour.